internal/transport/transport_test.go

```diff
@@ -2746,3 +2746,100 @@ func (s) TestClientSendsAGoAwayFrame(t *testing.T) {
+	// equivalent to network hang scenario when client is
+	// failing to write GOAWAY frame.
+	origGoAwayLoopyTimeout := goAwayLoopyWriterTimeout
+	goAwayLoopyWriterTimeout = 0
+	defer func() {
+		goAwayLoopyWriterTimeout = origGoAwayLoopyTimeout
+	}()
+
+	// Create a server.
+	lis, err := net.Listen("tcp", "localhost:0")
+	if err != nil {
+		t.Fatalf("Error while listening: %v", err)
+	}
+	defer lis.Close()
+	// serverGreetingDone is used to notify when server is done greeting the client.
+	serverGreetingDone := make(chan struct{})
+	// errorCh verifies that desired GOAWAY not received by server
+	errorCh := make(chan error, 1)
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	// Launch the server.
+	go func() {
+		defer close(errorCh)
+		conn, err := lis.Accept()
+		if err != nil {
+			t.Errorf("Error while accepting: %v", err)
+		}
+		defer conn.Close()
+		if _, err := io.ReadFull(conn, make([]byte, len(clientPreface))); err != nil {
+			t.Errorf("Error while reading client preface: %v", err)
+			return
+		}
+		framer := http2.NewFramer(conn, conn)
+		if err := framer.WriteSettings(); err != nil {
+			t.Errorf("Error while writing settings %v", err)
+			return
+		}
+		fr, _ := framer.ReadFrame()
+		if _, ok := fr.(*http2.SettingsFrame); !ok {
+			t.Errorf("Expected settings frame, got %T", fr)
+		}
+		fr, _ = framer.ReadFrame()
+		if fr, ok := fr.(*http2.SettingsFrame); !ok || !fr.IsAck() {
+			t.Errorf("Expected settings ACK frame, got %T", fr)
+		}
+		fr, _ = framer.ReadFrame()
+		if fr, ok := fr.(*http2.HeadersFrame); !ok || !fr.Flags.Has(http2.FlagHeadersEndHeaders) {
+			t.Errorf("Expected Headers frame with END_HEADERS frame, got %T", fr)
+		}
+		close(serverGreetingDone)
+
+		frame, err := framer.ReadFrame()
+		if err != nil {
+			return
+		}
+		switch fr := frame.(type) {
+		case *http2.GoAwayFrame:
+			if fr.ErrCode == http2.ErrCodeNo {
+				t.Logf("Received goAway frame from client")
+				return
+			}
+			errorCh <- fmt.Errorf("received unexpected goAway frame: %v", err)
+		default:
+			errorCh <- fmt.Errorf("server received a frame other than GOAWAY: %v", err)
+			return
+		}
+	}()
+
+	ct, err := NewClientTransport(ctx, ctx, resolver.Address{Addr: lis.Addr().String()}, ConnectOptions{}, func(GoAwayReason) {})
+	if err != nil {
+		t.Fatalf("Error while creating client transport: %v", err)
+	}
+	_, err = ct.NewStream(ctx, &CallHdr{})
+	if err != nil {
+		t.Fatalf("failed to open stream: %v", err)
+	}
+	// Wait until server receives the headers and settings frame as part of greet.
+	<-serverGreetingDone
+	// ct.Close will try to send the GOAWAY to server and will fail writing
+	// GOAWAY and will eventually close the loopyWriter as
+	// goAwayLoopyWriterTimeout (time out for writing GOAWAY) is set to zero, which is
+	// equivalent to network hang scenario.
+	ct.Close(errors.New("manually closed by client"))
+	select {
+	case <-errorCh:
+	case <-ctx.Done():
+		t.Errorf("Context timed out")
```

Comment:
Test error strings should give enough useful information for someone running the test and looking at the output. `Context timed out` does not tell me anything useful.